### PR TITLE
Kraken: refactor mvj

### PR DESCRIPTION
**Pull request by xlqian**
this is a first part of http://jira.canaltp.fr/browse/NAVP-358

The creation of vj is not yet implemented(see TODO)

### CODE REVIEW

File: source/kraken/realtime.cpp [hunk 1]
```diff
@@ -83,13 +63,18 @@ make_no_service_severity(const boost::posix_time::ptime& timestamp,
...
-    for (const auto* vj: mvj.base_vj) {
-        if (vj->base_validity_pattern()->check(date)) {
-            return execution_period(date, *vj);
+    boost::optional<boost::posix_time::time_period> res;
+    auto fun = [&](const nt::VehicleJourney& vj){
+        if (!res && vj.get_validity_pattern_at(type::RTLevel::Base)->check(date)) {
+            res = execution_period(date, vj);
         }
+    };
+    mvj.for_vjs_at_rt_level(type::RTLevel::Base, fun);
```
Comment (TeXitoi) [inline source/kraken/realtime.cpp]: You want to use `get_vjs_at_date` here.

File: source/type/rt_level.h [hunk 2]
```diff
@@ -39,6 +41,18 @@ enum class RTLevel : char {
     Adapted, // adapted schedule (planned maintenance, strike, ...)
     RealTime
 };
+
+inline RTLevel get_rt_level_from_string(const std::string& level_str) {
+    const static std::map<std::string, RTLevel> level2string_map = { {"theoric", RTLevel::Base},
```
Comment (TeXitoi) [inline source/type/rt_level.h:46]: using a map seems a bit overkill

File: source/type/type.cpp [hunk 3]
```diff
@@ -305,6 +306,56 @@ bool ValidityPattern::uncheck2(unsigned int day) const {
         return !days[day-1] && !days[day] && !days[day+1];
 }
 
+void MetaVehicleJourney::cancel_vj(RTLevel level,
+        const std::vector<boost::gregorian::date>& dates,
+        const transit_realtime::TripUpdate& /*trip_update*/,
```
Comment (TeXitoi) [inline source/type/type.cpp]: Why does this argument exist?

File: source/type/type.cpp [hunk 4]
```diff
@@ -305,6 +306,56 @@ bool ValidityPattern::uncheck2(unsigned int day) const {
...
+        const type::Data& data) {
+    auto func = [&](nt::VehicleJourney& vj){
+        for (const auto& date: dates){
+            // the train is canceled on one day, we just need to unset its realtime validitypattern
```
Comment (TeXitoi) [inline source/type/type.cpp]: comment unsyncronised

File: source/type/type.cpp [hunk 5]
```diff
@@ -305,6 +306,56 @@ bool ValidityPattern::uncheck2(unsigned int day) const {
...
+                    "canceling the vj " << vj.uri << " on " << date);
+            nt::ValidityPattern tmp_vp(*vj.get_validity_pattern_at(level));
+            tmp_vp.remove(date);
+            vj.validity_patterns[level] = data.pt_data->get_or_create_validity_pattern(tmp_vp);
+        }
+    };
+
+    for_vjs_at_rt_level(RTLevel::Base, func);
+    // we need to cancel all vj of at asked level
+    for_vjs_at_rt_level(level, func);
```
Comment (TeXitoi) [inline source/type/type.cpp]: No, the func must be called on every vjs. If you remove a realtime vj at a given date, but there is no rt vj this date, the adapted one will be the one valid, and we need to remove the date on it.

File: source/type/type.cpp [hunk 6]
```diff
@@ -305,6 +306,56 @@ bool ValidityPattern::uncheck2(unsigned int day) const {
...
+}
+
+std::vector<VehicleJourney*>
```
Comment (TeXitoi) [inline source/type/type.cpp]: This method can return 0 or 1 vj, thus I think `VehicleJourney*` is the type (and rename the method `get_vj_running_at_date`

File: source/type/type.cpp [hunk 7]
```diff
@@ -305,6 +306,56 @@ bool ValidityPattern::uncheck2(unsigned int day) const {
...
+MetaVehicleJourney::get_vjs_running_at_date(RTLevel level, const boost::gregorian::date& date){
+    std::vector<VehicleJourney*> res;
+
+    auto fun = [&](VehicleJourney& vj) {
+      if(vj.get_validity_pattern_at(level)->check(date)) {
```
Comment (TeXitoi) [inline source/type/type.cpp]: spacing.

File: source/type/type.cpp [hunk 8]
```diff
@@ -305,6 +306,56 @@ bool ValidityPattern::uncheck2(unsigned int day) const {
...
+          res.push_back(&vj);
+      }
+    };
+    for_vjs_at_rt_level(level, fun);
```
Comment (TeXitoi) [inline source/type/type.cpp]: No, it should be called on every vjs. If we ask for RT, but there is no RT, we must return the base one that is valid this date.

File: source/type/type.cpp [hunk 9]
```diff
@@ -305,6 +306,56 @@ bool ValidityPattern::uncheck2(unsigned int day) const {
...
+    return res;
+}
+
+std::vector<VehicleJourney*>
+MetaVehicleJourney::get_vjs_running_in_periode(RTLevel level, const boost::gregorian::date_period& period) {
```
Comment (TeXitoi) [inline source/type/type.cpp]: period doesn't have final `e` in english

File: source/type/type.cpp [hunk 10]
```diff
@@ -305,6 +306,56 @@ bool ValidityPattern::uncheck2(unsigned int day) const {
...
+    std::vector<VehicleJourney*> res;
+
+    auto day_itr = boost::gregorian::day_iterator{period.begin()};
+    for (; day_itr < period.end(); ++day_itr) {
+        auto running_vjs = get_vjs_running_at_date(level, *day_itr);
+        res.insert(res.end(),
+                std::make_move_iterator(running_vjs.begin()),
+                std::make_move_iterator(running_vjs.end()));
+    }
+    return res;
```
Comment (TeXitoi) [inline source/type/type.cpp:352]: This method is suboptimal, just looping on every vj, and checking all the date on cur_vj, and if valid, push_back and break.

File: source/type/type.h [hunk 11]
```diff
@@ -38,6 +38,7 @@ www.navitia.io
 #include "utils/exception.h"
 #include "utils/functions.h"
 #include "utils/idx_map.h"
+#include "type/gtfs-realtime.pb.h"
```
Comment (TeXitoi) [inline source/type/type.h]: I think it's a bad idea, and the implementation does not use TripUpdate...

File: source/type/type.h [hunk 12]
```diff
@@ -980,18 +984,63 @@ struct Calendar : public Nameable, public Header {
...
+    template<typename T>
+    void for_all_vjs(T fun) const{
+        for (const auto& rt_vjs: rtlevel_to_vjs_map) {
+            auto& vjs = rt_vjs.second;
+            boost::for_each(vjs, [&](VehicleJourney* vj){fun(*vj);});
+        }
+    }
+
+    template<typename T>
+    void for_vjs_at_rt_level(RTLevel level, T fun ) const{
```
Comment (TeXitoi) [inline source/type/type.h]: This method does not really have sens.

File: source/type/type.cpp (hunk 8 above)
Comment (xlqian) [inline source/type/type.cpp]: ok

File: source/type/rt_level.h (hunk 2 above)
Comment (xlqian) [inline source/type/rt_level.h:46]: what would you do?

File: source/kraken/realtime.cpp (hunk 1 above)
Comment (xlqian) [inline source/kraken/realtime.cpp]: done

File: source/type/type.cpp (hunk 5 above)
Comment (xlqian) [inline source/type/type.cpp]: done

File: source/kraken/realtime.cpp [hunk 13]
```diff
@@ -83,13 +63,12 @@ make_no_service_severity(const boost::posix_time::ptime& timestamp,
...
-        if (vj->base_validity_pattern()->check(date)) {
-            return execution_period(date, *vj);
-        }
+    auto running_vj = mvj.get_vj_at_date(type::RTLevel::Base, date);
+    if (running_vj) {
+        return execution_period(date, *running_vj);
     }
     // should be dead code
-    return execution_period(date, *mvj.base_vj.front());
+    return execution_period(date, *mvj.get_first_vj_at(type::RTLevel::Base));
```
Comment (antoine-de) [inline source/kraken/realtime.cpp]: Why not return a null period ?

File: source/type/type.h [hunk 14]
```diff
@@ -980,18 +984,62 @@ struct Calendar : public Nameable, public Header {
...
+        boost::for_each(rtlevel_to_vjs_map[level], [&](VehicleJourney* vj){fun(*vj);});
+    }
+
+    bool has_no_vjs_at(RTLevel level ) const {
```
Comment (antoine-de) [inline source/type/type.h]: I don't think this method is really usefull, this is used only in an assert, and does not have a real meaning apart from the BaseSchedule.

maybe we can remove the assert or give access to the BaseSchedule VJ ?

File: source/type/type.h [hunk 15]
```diff
@@ -980,18 +984,62 @@ struct Calendar : public Nameable, public Header {
...
+    // TODO XL: this function should be called in places where add_vj is called, because MetaVehicleJourney will
+    //          have the ownership of vjs instead of data
+    VehicleJourney* create_vj(/*args,*/ RTLevel level);
+
+    void add_vj(VehicleJourney* vj, RTLevel level) {
+        rtlevel_to_vjs_map[level].push_back(vj);
     }
+
+
+    VehicleJourney* get_first_vj_at(navitia::type::RTLevel level) const{
```
Comment (antoine-de) [inline source/type/type.h]: same here I don't think it has any meaning apart from the BaseSchedule

File: source/type/type.h (hunk 12 above)
Comment (antoine-de) [inline source/type/type.h]: Yes Maybe we can use the code in get_vj_at_date to do iether an iterator on vj\* from a RTLevel or a functor.

File: source/type/type.h (hunk 11 above)
Comment (xlqian) [inline source/type/type.h]: done

File: source/type/type.h (hunk 12 above)
Comment (xlqian) [inline source/type/type.h]: I'll return a range_iterator of vjs

File: source/type/type.h (hunk 15 above)
Comment (xlqian) [inline source/type/type.h]: ok

File: source/kraken/realtime.cpp (hunk 13 above)
Comment (xlqian) [inline source/kraken/realtime.cpp]: done

File: source/type/type.h (hunk 14 above)
Comment (xlqian) [inline source/type/type.h]: removed

File: source/type/rt_level.h [hunk 16]
```diff
@@ -30,15 +30,33 @@ www.navitia.io
...
 
 namespace type {
 enum class RTLevel : char {
-    Base = 0, // base schedule (theoretical, from the GTFS)
+    Begin = 0,
+    Base = Begin, // base schedule (theoretical, from the GTFS)
     Adapted, // adapted schedule (planned maintenance, strike, ...)
-    RealTime
+    RealTime,
+    End = RealTime
```
Comment (TeXitoi) [inline source/type/rt_level.h]: beark!

File: source/type/type.cpp [hunk 17]
```diff
@@ -305,6 +306,53 @@ bool ValidityPattern::uncheck2(unsigned int day) const {
         return !days[day-1] && !days[day] && !days[day+1];
 }
 
+void MetaVehicleJourney::cancel_vj(RTLevel level,
+        const std::vector<boost::gregorian::date>& dates,
+        const type::Data& data) {
+    for (auto l = level; l >= RTLevel::Begin; --l) {
+        for (auto* vj: rtlevel_to_vjs_map[l]) {
+            for (const auto& date: dates){
+                if (! vj->get_validity_pattern_at(level)->check(date)) { break; }
```
Comment (TeXitoi) [inline source/type/type.cpp]: You want a continue, no?

File: source/type/type.cpp [hunk 18]
```diff
@@ -305,6 +306,53 @@ bool ValidityPattern::uncheck2(unsigned int day) const {
...
+}
+
+std::vector<VehicleJourney*>
+MetaVehicleJourney::get_vjs_in_period(RTLevel level, const boost::gregorian::date_period& period) const{
+    std::vector<VehicleJourney*> res;
+    for (auto l = level; l >= RTLevel::Begin; --l) {
+        for (auto* vj: rtlevel_to_vjs_map[l]) {
+            auto day_itr = boost::gregorian::day_iterator{period.begin()};
+            for (; day_itr < period.end(); ++day_itr) {
+                if(vj->get_validity_pattern_at(l)->check(*day_itr)){
```
Comment (TeXitoi) [inline source/type/type.cpp:345]: you want level here, not `l`. you can just use for_all_vjs

File: source/type/type.cpp [hunk 19]
```diff
@@ -305,6 +306,53 @@ bool ValidityPattern::uncheck2(unsigned int day) const {
...
+            }
+        }
+    }
+}
+
+VehicleJourney*
+MetaVehicleJourney::get_vj_at_date(RTLevel level, const boost::gregorian::date& date) const{
+    for (auto l = level; l >= RTLevel::Begin; --l) {
+        for (auto* vj: rtlevel_to_vjs_map[l]) {
+            if(vj->get_validity_pattern_at(l)->check(date)){
```
Comment (TeXitoi) [inline source/type/type.cpp:330]: you want level here, not l

File: source/type/type.h [hunk 20]
```diff
@@ -980,18 +983,55 @@ struct Calendar : public Nameable, public Header {
...
+    VehicleJourney* create_vj(/*args,*/ RTLevel level);
+
+    void add_vj(VehicleJourney* vj, RTLevel level) {
+        rtlevel_to_vjs_map[level].push_back(vj);
+    }
+
+    template<typename T>
+    void for_all_vjs(T fun) const{
+        for (const auto& rt_vjs: rtlevel_to_vjs_map) {
+            auto& vjs = rt_vjs.second;
```
Comment (TeXitoi) [inline source/type/type.h:1007]: is this variable really useful?

File: source/type/type.h [hunk 21]
```diff
@@ -980,18 +983,55 @@ struct Calendar : public Nameable, public Header {
...
+        }
     }
+
+    template<typename T>
+    void for_vjs_at_rt_level(RTLevel level, T fun ) const{
+        boost::for_each(rtlevel_to_vjs_map[level], [&](VehicleJourney* vj){fun(*vj);});
+    }
+
+
+    using VjIter = boost::iterator_range<std::vector<VehicleJourney*>::const_iterator>;
```
Comment (TeXitoi) [inline source/type/type.h]: Why not just returning a const reference to the std::vector?

File: source/type/rt_level.h (hunk 16 above)
Comment (xlqian) [inline source/type/rt_level.h]: I didn't have many choices....

File: source/type/type.cpp (hunk 17 above)
Comment (xlqian) [inline source/type/type.cpp]: yep

File: source/type/type.h (hunk 20 above)
Comment (xlqian) [inline source/type/type.h:1007]: I just wanted to give a name to `rt_vjs.second;`

File: source/type/type.cpp (hunk 17 above)
Comment (xlqian) [inline source/type/type.cpp]: it is

**Comment by xlqian** [comment]
Jenkins, test this please

**Comment by xlqian** [comment]
Jenkins, test this please

**Comment by xlqian** [comment]
Jenkins, test this please

**Comment by Hoshiyo** [comment]
Jenkins, test this please

File: source/type/rt_level.h (hunk 2 above)
Comment (xlqian) [inline source/type/rt_level.h:46]: I'll  fix this in the next PR

**Pull request merged**